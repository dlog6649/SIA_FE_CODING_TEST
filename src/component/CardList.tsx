import React, { useEffect, useRef } from "react";
import { useHistory } from "react-router-dom";

const url = "https://jsonplaceholder.typicode.com/photos";
const proxyurl = "https://cors-anywhere.herokuapp.com/";

interface Props {
  viewImage: (url: string, title: string) => void;
}

export default function CardList({ viewImage }: Props) {
  const refCardList = useRef<HTMLDivElement>(null);
  const history = useHistory();

  useEffect(() => {
    console.log("CardList useEffect");
    fetch(proxyurl + url)
      .then((response) => response.json())
      .then((json) => {
        if (!refCardList.current) {
          return;
        }
        const cardList = refCardList.current;
        let cards = "";
        for (let i = 0; i < 8; i++) {
          cards += `
            <div class="card">
              <div class="card-thumbnail">
                  <img class="thumbnail" data-testid="testThumbnail" src="${json[i].thumbnailUrl}" data-url="${json[i].url}" data-title="${json[i].title}">
              </div>
              <div class="card-title" title="${json[i].title}">${json[i].title}</div>
            </div>
          `;
        }
        cardList.insertAdjacentHTML("afterbegin", cards);
      })
      .catch((error) => alert(`fetch failed\nerror: ${error}`));
  }, []);

  const viewImg = (evt: any) => {
    if (evt.target.className === "thumbnail") {
      viewImage(evt.target.dataset.url, evt.target.dataset.title);
    }
    history.push("/view");
  };

  return <div className="card-list" onClick={viewImg} data-testid="testCardList" ref={refCardList} />;
}
