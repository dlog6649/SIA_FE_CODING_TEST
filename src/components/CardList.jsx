import React, {useEffect} from 'react';

export default function CardList(props) {
    
    
    useEffect(() => {
        console.log('CardList useEffect');

        fetch('https://jsonplaceholder.typicode.com/photos')
            .then(response => response.json())
            .then(json => {
                const cardList = document.querySelector('.card-list');
                let cards = '';
                for(let i = 0; i < 8; i++) {
                    cards += `
                        <div class="card">
                            <div class="card-thumbnail">
                                <img class="thumbnail" data-testid="testThumbnail" src="${json[i].thumbnailUrl}" data-url="${json[i].url}" data-title="${json[i].title}">
                            </div>
                            <div class="card-title" title="${json[i].title}">${json[i].title}</div>
                        </div>
                    `;
                }
                cardList.insertAdjacentHTML('afterbegin', cards);
            })
            .catch(error => alert('fetch failed\nerror: ' + error));
    },[]);


    const viewImg = e => {
        if(e.target.className === 'thumbnail') {
            props.viewImage(e.target.dataset.url, e.target.dataset.title);
        }
    };


    return (
        <div>
            <div className="card-list" onClick={viewImg} data-testid="testCardList">
            </div>
        </div>
    );
}