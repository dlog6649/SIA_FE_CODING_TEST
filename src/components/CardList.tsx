import React, { useEffect, useRef } from 'react';

const url: string = 'https://jsonplaceholder.typicode.com/photos';
const proxyurl: string = "https://cors-anywhere.herokuapp.com/";

interface Props {
    viewImage: any
}

export default function CardList(props: Props) {
    let refCardList = useRef(null);
    useEffect(() => {
        console.log('CardList useEffect');
        fetch(proxyurl + url)
            .then((response) => response.json())
            .then((json) => {
                const cardList: any = refCardList.current;
                let cards: string = '';
                for(let i = 0; i < 8; i++) {
                    cards += `
                        <div class="card">
                            <div class="card-thumbnail">
                                <img class="thumbnail" data-testid="testThumbnail" src="${json[i].thumbnailUrl}" data-url="${json[i].url}" data-title="${json[i].title}">
                            </div>
                            <div class="card-title" title="${json[i].title}">${json[i].title}</div>
                        </div>
                    `;
                }
                cardList.insertAdjacentHTML('afterbegin', cards);
            })
            .catch(error => alert('fetch failed\nerror: ' + error));
    },[]);

    const viewImg = (e: any) => {
        if(e.target.className === 'thumbnail') {
            props.viewImage(e.target.dataset.url, e.target.dataset.title);
        }
    };

    return (
        <div>
            <div className="card-list" onClick={viewImg} data-testid="testCardList" ref={refCardList} >
            </div>
        </div>
    );
}