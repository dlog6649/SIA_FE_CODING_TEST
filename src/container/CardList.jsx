import CardList from '../components/CardList';
import { connect } from 'react-redux';
import { viewImage } from '../modules/annotator';
import { push } from 'connected-react-router';


const mapDispatchToProps = dispatch => {
    return {
        viewImage:(url, title) => {
            dispatch(viewImage(url, title));
            dispatch(push('/view'));
        }
    };
};


export default connect(null, mapDispatchToProps)(CardList);