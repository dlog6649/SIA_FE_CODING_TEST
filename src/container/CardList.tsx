import CardList from '../components/CardList';
import { connect } from 'react-redux';
import { viewImage } from '../modules/annotator';
import { push } from 'connected-react-router';

const mapDispatchToProps = (dispatch: any) => {
    return {
        viewImage:(url: string, title: string) => {
            dispatch(viewImage(url, title));
            dispatch(push('/view'));
        }
    };
};

export default connect(null, mapDispatchToProps)(CardList);