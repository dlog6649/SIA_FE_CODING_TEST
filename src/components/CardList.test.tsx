import React from 'react';
import { render, fireEvent, waitForElement } from "@testing-library/react";
import CardList from './CardList';


describe('CardList 테스트 시작', () => {
  let wrapper: any = null;
  const mockViewImage = jest.fn();
  jest.setTimeout(30000);

  it('렌더링된 스냅샷이 기존과 일치해야 한다. 그리고 썸네일을 클릭 시 해당 정보가 디스패치돼야 한다.', async () => {
    wrapper = render(<CardList viewImage={mockViewImage} />);

    // jsonplaceholder 비동기 대기시간으로 30초 부여
    await waitForElement(() => wrapper.getAllByTestId('testThumbnail'), {timeout: 30000})
      .then(() => {
        expect(wrapper.baseElement).toMatchSnapshot();
        
        let testThumbnails = wrapper.getAllByTestId('testThumbnail');
        expect(testThumbnails.length).toBe(8);
        
        // 0번 썸네일 클릭
        fireEvent.click(testThumbnails[0]);

        let viewImageEventByClick = mockViewImage.mock.calls[0];
        let [url, title] = viewImageEventByClick;
        expect(url).toEqual('https://via.placeholder.com/600/92c952');
        expect(title).toEqual('accusamus beatae ad facilis cum similique qui sunt');
      })
      .catch(err => console.log(err));
  });
});
