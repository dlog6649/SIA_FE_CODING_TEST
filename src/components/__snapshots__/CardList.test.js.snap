// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardList 테스트 시작 렌더링된 스냅샷이 기존과 일치해야 한다. 그리고 썸네일을 클릭 시 해당 정보가 디스패치돼야 한다. 1`] = `
<body>
  <div>
    <div>
      <div
        class="card-list"
        data-testid="testCardList"
      >
        
                        
        <div
          class="card"
        >
          
                            
          <div
            class="card-thumbnail"
          >
            
                                
            <img
              class="thumbnail"
              data-testid="testThumbnail"
              data-title="accusamus beatae ad facilis cum similique qui sunt"
              data-url="https://via.placeholder.com/600/92c952"
              src="https://via.placeholder.com/150/92c952"
            />
            
                            
          </div>
          
                            
          <div
            class="card-title"
            title="accusamus beatae ad facilis cum similique qui sunt"
          >
            accusamus beatae ad facilis cum similique qui sunt
          </div>
          
                        
        </div>
        
                    
                        
        <div
          class="card"
        >
          
                            
          <div
            class="card-thumbnail"
          >
            
                                
            <img
              class="thumbnail"
              data-testid="testThumbnail"
              data-title="reprehenderit est deserunt velit ipsam"
              data-url="https://via.placeholder.com/600/771796"
              src="https://via.placeholder.com/150/771796"
            />
            
                            
          </div>
          
                            
          <div
            class="card-title"
            title="reprehenderit est deserunt velit ipsam"
          >
            reprehenderit est deserunt velit ipsam
          </div>
          
                        
        </div>
        
                    
                        
        <div
          class="card"
        >
          
                            
          <div
            class="card-thumbnail"
          >
            
                                
            <img
              class="thumbnail"
              data-testid="testThumbnail"
              data-title="officia porro iure quia iusto qui ipsa ut modi"
              data-url="https://via.placeholder.com/600/24f355"
              src="https://via.placeholder.com/150/24f355"
            />
            
                            
          </div>
          
                            
          <div
            class="card-title"
            title="officia porro iure quia iusto qui ipsa ut modi"
          >
            officia porro iure quia iusto qui ipsa ut modi
          </div>
          
                        
        </div>
        
                    
                        
        <div
          class="card"
        >
          
                            
          <div
            class="card-thumbnail"
          >
            
                                
            <img
              class="thumbnail"
              data-testid="testThumbnail"
              data-title="culpa odio esse rerum omnis laboriosam voluptate repudiandae"
              data-url="https://via.placeholder.com/600/d32776"
              src="https://via.placeholder.com/150/d32776"
            />
            
                            
          </div>
          
                            
          <div
            class="card-title"
            title="culpa odio esse rerum omnis laboriosam voluptate repudiandae"
          >
            culpa odio esse rerum omnis laboriosam voluptate repudiandae
          </div>
          
                        
        </div>
        
                    
                        
        <div
          class="card"
        >
          
                            
          <div
            class="card-thumbnail"
          >
            
                                
            <img
              class="thumbnail"
              data-testid="testThumbnail"
              data-title="natus nisi omnis corporis facere molestiae rerum in"
              data-url="https://via.placeholder.com/600/f66b97"
              src="https://via.placeholder.com/150/f66b97"
            />
            
                            
          </div>
          
                            
          <div
            class="card-title"
            title="natus nisi omnis corporis facere molestiae rerum in"
          >
            natus nisi omnis corporis facere molestiae rerum in
          </div>
          
                        
        </div>
        
                    
                        
        <div
          class="card"
        >
          
                            
          <div
            class="card-thumbnail"
          >
            
                                
            <img
              class="thumbnail"
              data-testid="testThumbnail"
              data-title="accusamus ea aliquid et amet sequi nemo"
              data-url="https://via.placeholder.com/600/56a8c2"
              src="https://via.placeholder.com/150/56a8c2"
            />
            
                            
          </div>
          
                            
          <div
            class="card-title"
            title="accusamus ea aliquid et amet sequi nemo"
          >
            accusamus ea aliquid et amet sequi nemo
          </div>
          
                        
        </div>
        
                    
                        
        <div
          class="card"
        >
          
                            
          <div
            class="card-thumbnail"
          >
            
                                
            <img
              class="thumbnail"
              data-testid="testThumbnail"
              data-title="officia delectus consequatur vero aut veniam explicabo molestias"
              data-url="https://via.placeholder.com/600/b0f7cc"
              src="https://via.placeholder.com/150/b0f7cc"
            />
            
                            
          </div>
          
                            
          <div
            class="card-title"
            title="officia delectus consequatur vero aut veniam explicabo molestias"
          >
            officia delectus consequatur vero aut veniam explicabo molestias
          </div>
          
                        
        </div>
        
                    
                        
        <div
          class="card"
        >
          
                            
          <div
            class="card-thumbnail"
          >
            
                                
            <img
              class="thumbnail"
              data-testid="testThumbnail"
              data-title="aut porro officiis laborum odit ea laudantium corporis"
              data-url="https://via.placeholder.com/600/54176f"
              src="https://via.placeholder.com/150/54176f"
            />
            
                            
          </div>
          
                            
          <div
            class="card-title"
            title="aut porro officiis laborum odit ea laudantium corporis"
          >
            aut porro officiis laborum odit ea laudantium corporis
          </div>
          
                        
        </div>
        
                    
      </div>
    </div>
  </div>
</body>
`;
